python/cuml/ensemble/randomforest_shared.pyx

```diff
@@ -0,0 +1,131 @@
+        void* buf
+        char* format
+        size_t itemsize
+        size_t nitem
+    cdef cppclass Model:
+        vector[PyBufferFrame] GetPyBuffer() except +
+        void InitFromPyBuffer(vector[PyBufferFrame] frames) except +
+
+cdef extern from "Python.h":
+    Py_buffer* PyMemoryView_GET_BUFFER(PyObject* mview)
+
+cdef class PyBufferFrameWrapper:
+    cdef PyBufferFrame _handle
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        pass
+
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        cdef Py_ssize_t itemsize = self._handle.itemsize
+
+        self.shape[0] = self._handle.nitem
+        self.strides[0] = itemsize
+
+        buffer.buf = self._handle.buf
+        buffer.format = self._handle.format
+        buffer.internal = NULL
+        buffer.itemsize = itemsize
+        buffer.len = self._handle.nitem * itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self.shape
+        buffer.strides = self.strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+cdef PyBufferFrameWrapper MakePyBufferFrameWrapper(PyBufferFrame handle):
+    cdef PyBufferFrameWrapper wrapper = PyBufferFrameWrapper()
+    wrapper._handle = handle
+    return wrapper
+
+cdef list _get_frames(ModelHandle model):
+    frames = []
+    cdef Model* model_obj = <Model*>model
+    cdef vector[PyBufferFrame] interface = model_obj.GetPyBuffer()
+    cdef vector[PyBufferFrame].iterator it = interface.begin()
+    while it != interface.end():
+        v = deref(it)
+        w = MakePyBufferFrameWrapper(v)
+        frames.append(memoryview(w))
+        inc(it)
+    return frames
+
+cdef ModelHandle _init_from_frames(vector[PyBufferFrame] frames) except *:
+    cdef Model* model_obj = new Model()
+    model_obj.InitFromPyBuffer(frames)
+    cdef ModelHandle handle = <ModelHandle>model_obj
+    return handle
+
+
+def get_frames(model: uintptr_t) -> List[memoryview]:
+    return _get_frames(<ModelHandle> model)
+
+
+def init_from_frames(frames: List[np.ndarray],
+                     format_str: List[str], itemsize: List[int]) -> uintptr_t:
+    cdef vector[PyBufferFrame] cpp_frames
+    cdef Py_buffer* buf
+    cdef PyBufferFrame cpp_frame
+    for i, frame in enumerate(frames):
+        x = memoryview(frame)
+        buf = PyMemoryView_GET_BUFFER(<PyObject*>x)
+        cpp_frame.buf = buf.buf
+        cpp_frame.format = format_str[i]
+        cpp_frame.itemsize = itemsize[i]
+        cpp_frame.nitem = buf.len // itemsize[i]
+        cpp_frames.push_back(cpp_frame)
+    return <uintptr_t> _init_from_frames(cpp_frames)
+
+
+def treelite_serialize(
+    model: uintptr_t
+) -> Dict[str, Union[List[str], List[np.ndarray]]]:
+    frames = get_frames(model)
+    header = {'format_str': [x.format.encode('utf-8') for x in frames],
```

Comment:
Keeping `encode('utf-8')`, because `.view()` is untenable. See my comment above.

Reply:
Maybe we can handle the `encode` as part of `init_from_frames`?

Reply:
Can you explain why we should move `encode` to `init_from_frames`? To me, the string encoding is a proper part of the serialization process.

Reply:
Also, we can only assign bytes to the variable `cpp_frame.format` in `init_from_frames`. Trying to assign `format_str[i].encode()` is an error because we cannot use a Python temporary in assigning to C fields.

Reply:
Sorry maybe this clearer if all related changes are visible. Have grouped them in this review ( https://github.com/rapidsai/cuml/pull/2263#pullrequestreview-431932349 ).